internal/statestore/backfill.go

```diff
@@ -0,0 +1,152 @@
+package statestore
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/gomodule/redigo/redis"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"open-match.dev/open-match/pkg/pb"
+)
+
+// CreateBackfill creates a new Backfill in the state storage. If the id already exists, it will be overwritten.
+func (rb *redisBackend) CreateBackfill(ctx context.Context, backfill *pb.Backfill) error {
+	redisConn, err := rb.redisPool.GetContext(ctx)
+	if err != nil {
+		return status.Errorf(codes.Unavailable, "CreateBackfill, id: %s, failed to connect to redis: %v", backfill.GetId(), err)
+	}
+	defer handleConnectionClose(&redisConn)
+
+	value, err := proto.Marshal(backfill)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to marshal the backfill proto, id: %s", backfill.GetId())
+		return status.Errorf(codes.Internal, "%v", err)
+	}
+
+	_, err = redisConn.Do("SET", backfill.GetId(), value)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to set the value for backfill, id: %s", backfill.GetId())
+		return status.Errorf(codes.Internal, "%v", err)
+	}
+
+	return nil
+}
+
+// GetBackfill gets the Backfill with the specified id from state storage. This method fails if the Backfill does not exist.
+func (rb *redisBackend) GetBackfill(ctx context.Context, id string) (*pb.Backfill, error) {
+	redisConn, err := rb.redisPool.GetContext(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Unavailable, "GetBackfill, id: %s, failed to connect to redis: %v", id, err)
+	}
+	defer handleConnectionClose(&redisConn)
+
+	return getBackfill(redisConn, id)
+}
+
+// DeleteBackfill removes the Backfill with the specified id from state storage. This method succeeds if the Backfill does not exist.
+func (rb *redisBackend) DeleteBackfill(ctx context.Context, id string) error {
+	redisConn, err := rb.redisPool.GetContext(ctx)
+	if err != nil {
+		return status.Errorf(codes.Unavailable, "DeleteBackfill, id: %s, failed to connect to redis: %v", id, err)
+	}
+	defer handleConnectionClose(&redisConn)
+
+	_, err = redisConn.Do("DEL", id)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to delete the backfill from state storage, id: %s", id)
+		return status.Errorf(codes.Internal, "%v", err)
+	}
+
+	return nil
+}
+
+// UpdateBackfill updates an existing Backfill with new data.
+func (rb *redisBackend) UpdateBackfill(ctx context.Context, backfill *pb.Backfill, updateFunc func(current *pb.Backfill, new *pb.Backfill) (*pb.Backfill, error)) (*pb.Backfill, error) {
```

Comment:
It's not clear to me where this update function will be used?

This logic seems to be an optimistic update.
If the backfill is updated by the MMF, as soon as the backfill is updated by anything else there's no reason to retry - the whole fetch match cycle is optimistic and has failed.
When updating from the gameserver, it needs to forcefully update and set the value.

When acknowledging from the gameserver, this field isn't actually updated?
How are the non-user visible fields (pending tickets, last acknowledge time) going to work with this?

Reply:
> When acknowledging from the gameserver, this field isn't actually updated?

Cold you please clarify which field are you talking about? 

> How are the non-user visible fields (pending tickets, last acknowledge time) going to work with this?

We expect last acknowledge time to be an in-memory variable. 
Regarding pending tickets - next PR will introduce methods to store _ticketIDs_ by _backfillID-generation_ key.

Reply:
I've added an isGS flag to this update function. Is it better?

Reply:
> Could you please clarify which field are you talking about?
The redis field, the whole backfill ticket has no changes.

> We expect last acknowledge time to be an in-memory variable.
How does this work?

> Regarding pending tickets - next PR will introduce methods to store ticketIDs by backfillID-generation key.
Hm, that might work, but I'm not quite seeing it.

@aLekSer @akremsa Would you please update the design doc with the fields which are used in redis and the general strategy for working with them?  I think that would help avoid churn on code review.

Reply:
Yes, I think this should be clear in the design doc, updating it.

> When acknowledging from the gameserver, this field isn't actually updated?

On acknowledgment we expect not to make any updates to Redis, only read from Redis and return back to the GameServer the current state.

>How does this work?

There is no acknowledge time, so with every `acknowledge` request it would be stored in an LRU cache or in a simple `map[backfillId]timestamp`. It was mentioned in discussion that we should not propagate this timestamp to the database, because it would cause frequent updates.

Reply:
May be this is a good idea or not but can we add simple methods to state store like:

1. UpdateBackfill(backfill) - just updates backfill value by key
2. UpdateBackfillTickets(backfill, ticketIds) - just update list of tickets associated with backfill. Stored in Redis like [backfillId_tickets] -> [ticket ids]
3. Lock - distributed lock
4. Unlock - distributed unlock

To backend:

1. UpdateBackfill(backfill)

- lock
- get backfill
- check if exists
- release associated tickets
- increment generation
- update backfill
- unlock

2. SysUpdateBackfill(backfill, ticketIds)
- lock
- get backfill
- check if exists
- compare generation
- update backfill
- update associated tickets
- unlock

Reply:
On AcknowledgeBackfill:

- lock
- get backfill
- check if exists
- assign associated tickets
- unlock


Return Backfill without any tickets.

Reply:
release associated tickets - remove tickets from pending state. Now they are available for QueryTickets.

Reply:
Storing acknowledge timestamps in the frontend or backends won't work, as there are multiple instances of those services.

Lock/unlock shouldn't be visible from the app code for frontend/backend/etc, as the requirement to do locking is because of our use of redis ( I think? ), so I think it would be better to do more meaningful methods inside statestore, which may use private methods to locking, etc.